### PR TITLE
Support Other Parachain <> Darwinia <> Ethereum

### LINKED ISSUE #1062: Connect Ethereum <> Other Parachains
## If we keep Darwinia as a solo chain.
Protocol: XCMP <> router <> LCMP
Path: Other Parachains(on Polkadot) <> Darwinia Parachain <> Darwinia <> Ethereum 
Path: Other Parachains(on Kusama) <> Crab Parachain <> Crab <> Darwinia <> Ethereum
## If Darwinia becomes a parachain.
Convert solo chain to parachain: **_requires verification_**
https://docs.substrate.io/reference/how-to-guides/parachains/convert-a-solo-chain/
Protocol: XCMP <> LCMP
Path: Other Parachains <> Darwinia Parachain <> Ethereum

### PR DESCRIPTION
**Pull request by jiguantong**
Closes #1062 #973 
## Develop
- [x] Add pallet-ethereum-xcm
- [x] Add pallet-xcm-transactor
- [x] Block production too long
- [x] Moonbeam 0.9.38
- [x] Astar XCM v3
- [x] DescendOrigin Barrier
- [x] Goerli Endpoint @wuminzhe 
- [x] Darwinia Endpoint @wuminzhe 

### Astar > Goerli
- [x] Pangolin2 XCM > EVM
- [x] Astar XCM > Pangolin2 EVM
- [x] Astar XCM > Goerli deliver
- [x] Astar XCM > Goerli execution

### Goerli > Astar
- [x] Goerli > Pangolin2 EVM
- [x] Pangolin2 EVM XCM > Astar
- [x] Goerli > Astar

### Finishing work
- [x] Add related pallets to darwinia-runtime
- [x] Use purestake/moonbeam v0.9.38 https://github.com/PureStake/moonbeam/pull/2145
- [x] Remove debug patch
- [x] Clean debug & unused code
- [x] Pick #1022

### CODE REVIEW

**Review by boundless-forest — APPROVED** [review]
I mainly review things related to the EVM module. LGTM

File: pallet/message-transact/Cargo.toml [hunk 1]
```diff
@@ -46,6 +46,7 @@ sp-runtime       = { workspace = true, features = ["std"] }
 
 [features]
 default = ["std"]
+
```
Comment (AurevoirXavier) [inline pallet/message-transact/Cargo.toml:49]: Remove this empty line.

File: runtime/darwinia/Cargo.toml [hunk 2]
```diff
@@ -70,8 +70,9 @@ pallet-evm-precompile-modexp   = { workspace = true }
 pallet-evm-precompile-simple   = { workspace = true }
 
 # moonbeam
-precompile-utils = { workspace = true }
-xcm-primitives   = { workspace = true }
+pallet-ethereum-xcm    = { workspace = true }
```
Comment (AurevoirXavier) [inline runtime/darwinia/Cargo.toml:73]: Extra empty spaces.

File: runtime/darwinia/src/lib.rs [hunk 3]
```diff
@@ -163,7 +163,10 @@ frame_support::construct_runtime! {
 		BridgeKusamaParachain: pallet_bridge_parachains::<Instance1> = 40,
 		BridgeCrabMessages: pallet_bridge_messages::<Instance1> = 41,
 		BridgeCrabDispatch: pallet_bridge_dispatch::<Instance1> = 42,
-		CrabFeeMarket: pallet_fee_market::<Instance1> = 43
+		CrabFeeMarket: pallet_fee_market::<Instance1> = 43,
+
+		// Ethereum XCM
+		EthereumXcm: pallet_ethereum_xcm = 44
```
Comment (AurevoirXavier) [inline runtime/darwinia/src/lib.rs:169]: Suggest move to `// EVM stuff`

File: runtime/pangoro/Cargo.toml [hunk 4]
```diff
@@ -70,8 +70,9 @@ pallet-evm-precompile-modexp   = { workspace = true }
 pallet-evm-precompile-simple   = { workspace = true }
 
 # moonbeam
-precompile-utils = { workspace = true }
-xcm-primitives   = { workspace = true }
+pallet-ethereum-xcm    = { workspace = true }
```
Comment (AurevoirXavier) [inline runtime/pangoro/Cargo.toml:73]: Format.

File: runtime/pangoro/src/lib.rs [hunk 5]
```diff
@@ -165,7 +165,10 @@ frame_support::construct_runtime! {
 		BridgeRococoParachain: pallet_bridge_parachains::<Instance1> = 40,
 		BridgePangolinMessages: pallet_bridge_messages::<Instance1> = 41,
 		BridgePangolinDispatch: pallet_bridge_dispatch::<Instance1> = 42,
-		PangolinFeeMarket: pallet_fee_market::<Instance1> = 43
+		PangolinFeeMarket: pallet_fee_market::<Instance1> = 43,
+
+		// Ethereum XCM
+		EthereumXcm: pallet_ethereum_xcm = 44,
```
Comment (AurevoirXavier) [inline runtime/pangoro/src/lib.rs:171]: Format

File: runtime/darwinia/src/pallets/polkadot_xcm.rs [hunk 6]
```diff
@@ -215,3 +221,63 @@ impl cumulus_pallet_xcm::Config for Runtime {
 	type RuntimeEvent = RuntimeEvent;
 	type XcmExecutor = xcm_executor::XcmExecutor<XcmExecutorConfig>;
 }
+
+pub struct EthereumXcmEnsureProxy;
+impl xcm_primitives::EnsureProxy<AccountId> for EthereumXcmEnsureProxy {
+	fn ensure_ok(delegator: AccountId, delegatee: AccountId) -> Result<(), &'static str> {
+		// The EVM implicitely contains an Any proxy, so we only allow for "Any" proxies
```
Comment (AurevoirXavier) [inline runtime/darwinia/src/pallets/polkadot_xcm.rs:228]: Could you explain this?

File: runtime/darwinia/src/pallets/polkadot_xcm.rs [hunk 7]
```diff
@@ -215,3 +221,63 @@ impl cumulus_pallet_xcm::Config for Runtime {
...
+		let def: pallet_proxy::ProxyDefinition<AccountId, pallets::proxy::ProxyType, BlockNumber> =
+			pallet_proxy::Pallet::<Runtime>::find_proxy(
+				&delegator,
+				&delegatee,
+				Some(pallets::proxy::ProxyType::Any),
+			)
+			.map_err(|_| "proxy error: expected `ProxyType::Any`")?;
+		// We only allow to use it for delay zero proxies, as the call will immediatly be executed
+		frame_support::ensure!(def.delay.is_zero(), "proxy delay is Non-zero`");
```
Comment (AurevoirXavier) [inline runtime/darwinia/src/pallets/polkadot_xcm.rs:237]: And this.

File: runtime/darwinia/src/pallets/polkadot_xcm.rs [hunk 8]
```diff
@@ -215,3 +221,63 @@ impl cumulus_pallet_xcm::Config for Runtime {
...
+	type ControllerOrigin = frame_system::EnsureRoot<AccountId>;
+	type EnsureProxy = EthereumXcmEnsureProxy;
+	type InvalidEvmTransactionError = pallet_ethereum::InvalidTransactionWrapper;
+	type ReservedXcmpWeight =
+		<Runtime as cumulus_pallet_parachain_system::Config>::ReservedXcmpWeight;
+	type ValidatedTransaction = pallet_ethereum::ValidatedTransaction<Self>;
+	type XcmEthereumOrigin = pallet_ethereum_xcm::EnsureXcmEthereumTransaction;
+}
+
+pub struct DarwiniaCall;
```
Comment (AurevoirXavier) [inline runtime/darwinia/src/pallets/polkadot_xcm.rs:252]: `DarwiniaCall`? This looks like a filter to me.

Comment (jiguantong) [reply runtime/darwinia/src/pallets/polkadot_xcm.rs:252]: Yes, it wraps Origin for ethereum_xcm transactions, making it equal to [`pallet_ethereum_xcm::Config::XcmEthereumOrigin`](https://github.com/darwinia-network/darwinia/blob/guantong/xcm-evm/runtime/darwinia/src/pallets/polkadot_xcm.rs#L249), and will be checked when [executing ethereum_xcm transactions](https://github.com/PureStake/moonbeam/blob/master/pallets/ethereum-xcm/src/lib.rs#L152)

File: runtime/darwinia/src/pallets/polkadot_xcm.rs (hunk 6 above)
Comment (jiguantong) [reply runtime/darwinia/src/pallets/polkadot_xcm.rs:228]: This is copied from moonbeam, I don't really understand it, but it is used in [`transact_through_proxy`](https://github.com/PureStake/moonbeam/blob/master/pallets/ethereum-xcm/src/lib.rs#L193), we don't need it, I will disable this.

File: runtime/darwinia/src/pallets/polkadot_xcm.rs [hunk 9]
```diff
@@ -111,6 +109,11 @@ frame_support::parameter_types! {
 	pub UniversalLocation: InteriorMultiLocation = Parachain(ParachainInfo::parachain_id().into()).into();
 	// One XCM operation is 1_000_000_000 weight - almost certainly a conservative estimate.
 	pub UnitWeightCost: frame_support::weights::Weight = frame_support::weights::Weight::from_parts(1_000_000_000, 64 * 1024);
+	pub LocalAssetsPalletLocation: MultiLocation = MultiLocation::new(
```
Comment (AurevoirXavier) [inline runtime/darwinia/src/pallets/polkadot_xcm.rs:112]: Unused.

File: runtime/darwinia/src/lib.rs [hunk 10]
```diff
@@ -156,6 +156,7 @@ frame_support::construct_runtime! {
 
 		// EVM stuff.
 		Ethereum: pallet_ethereum = 36,
+		EthereumXcm: pallet_ethereum_xcm = 44,
```
Comment (AurevoirXavier) [inline runtime/darwinia/src/lib.rs:159]: `XCM`, not `EVM`. Put it after the `CumulusXcm `.

File: runtime/pangolin/src/lib.rs [hunk 11]
```diff
@@ -157,6 +157,7 @@ frame_support::construct_runtime! {
 
 		// EVM stuff.
 		Ethereum: pallet_ethereum = 36,
+		EthereumXcm: pallet_ethereum_xcm = 44,
```
Comment (AurevoirXavier) [inline runtime/pangolin/src/lib.rs:160]: Same

File: runtime/pangolin/src/pallets/polkadot_xcm.rs [hunk 12]
```diff
@@ -111,6 +109,11 @@ frame_support::parameter_types! {
 	pub UniversalLocation: InteriorMultiLocation = Parachain(ParachainInfo::parachain_id().into()).into();
 	// One XCM operation is 1_000_000_000 weight - almost certainly a conservative estimate.
 	pub UnitWeightCost: frame_support::weights::Weight = frame_support::weights::Weight::from_parts(1_000_000_000, 64 * 1024);
+	pub LocalAssetsPalletLocation: MultiLocation = MultiLocation::new(
```
Comment (AurevoirXavier) [inline runtime/pangolin/src/pallets/polkadot_xcm.rs:112]: Unused

File: runtime/pangoro/src/lib.rs [hunk 13]
```diff
@@ -157,6 +157,7 @@ frame_support::construct_runtime! {
 
 		// EVM stuff.
 		Ethereum: pallet_ethereum = 36,
+		EthereumXcm: pallet_ethereum_xcm = 44,
```
Comment (AurevoirXavier) [inline runtime/pangoro/src/lib.rs:160]: Same

File: runtime/darwinia/src/pallets/polkadot_xcm.rs (hunk 6 above)
Comment (AurevoirXavier) [reply runtime/darwinia/src/pallets/polkadot_xcm.rs]: @boundless-forest help review the `EthereumXcmEnsureProxy`.

File: runtime/darwinia/src/pallets/polkadot_xcm.rs (hunk 8 above)
Comment (AurevoirXavier) [reply runtime/darwinia/src/pallets/polkadot_xcm.rs:241]: Might need a better name for this.

Comment (jiguantong) [reply runtime/darwinia/src/pallets/polkadot_xcm.rs:241]: `XcmCallFilter` ?

**Comment by AurevoirXavier** [comment]
Should we add those things for Crab? This can be an issue.

**Pull request merged**